lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -961,7 +961,29 @@ public function set_attribute( $name, $value ) {
 		if ( true === $value ) {
 			$updated_attribute = $name;
 		} else {
-			// @TODO: What escaping and sanitization do we need here?
+			/**
+			 * Prevents HTML injection attacks by encoding the " characters to &quot;
+			 *
+			 * This protects from values terminating the attribute value as follows:
+			 *
+			 * <code>
+			 *     $p = new WP_HTML_Tag_Processor( '<div class="header"></div>' );
+			 *     $p->next_tag();
+			 *     $p->set_attribute('class', '" onclick="alert');
+			 *     echo $p;
+			 *     // Danger! Not encoding the double quotes would create another HTML attribute:
+			 *     //    <div class="" onclick="alert"></div>
+			 *     // Fortunately the encoding is in place:
+			 *     //    <div class="&quot; onclick=&quot;alert"></div>
+			 * </code>
+			 *
+			 * This is sufficient because:
+			 * * The attributes touched by set_attribute are always rewritten using double quotes.
+			 * * The only way to terminate a double quoted attribute value is via a double quote (")
+			 *   character. There is no need to encode other characters such as <, >, /, or '.
```

Comment:
Given the existing body of code attempting to parse HTML in WordPress and its plugins, I think @azaozz has a good point that we might want to encode `<`, `>`, and `&` as well. This isn't an HTML thing, but if we allow `alt="the <img /> tag` we might open the door for defects in other code to trip up here when we had the power to prevent it.

the one caveat I see is with `class` because we don't support decoding character references at this moment, but even then I think it's fine to _encode_ them on save.

Reply:
Good point, let’s escape these as well.

Reply:
Actually, let's stick to encoding just `"`, `<`, and `>`. Encoding `&` breaks the following use-case:

```php
$p->set_attribute( 'data-price', '10 &euro;' );
// <div data-price="10 &amp;euro;"></div>
```

Reply:
I believe that `_wp_specialchars()` exists to handle encoding without double-encoding (unless you explicitly tell it to). it looks up any character references before replacing `&` willy-nilly.

```php
"10 &euro; &amp; &quot;USD&quot;" === _wp_specialchars( '10 &euro; & "USD"' );
```

Reply:
`_wp_specialchars` seems to be deprecated in favor of `esc_html`, which does the same thing as `esc_attr` – let's try `esc_attr` then?

Reply:
Proposed in 6c309a837c